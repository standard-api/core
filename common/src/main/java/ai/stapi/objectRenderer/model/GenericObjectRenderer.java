package ai.stapi.objectRenderer.model;

import ai.stapi.objectRenderer.exceptions.OptionsAreNotSupportedByAnyRendererException;
import ai.stapi.objectRenderer.exceptions.OptionsAreSupportedByMultipleRenderersException;
import java.util.List;

public class GenericObjectRenderer {

  private List<ObjectRenderer> existingObjectRenderers;

  public GenericObjectRenderer(List<ObjectRenderer> existingObjectRenderers) {
    this.existingObjectRenderers = existingObjectRenderers;
  }

  public RenderOutput render(Object obj, RendererOptions options) {
    var renderer = this.getObjectRenderer(options);
    return renderer.render(obj, options);
  }

  private ObjectRenderer getObjectRenderer(RendererOptions options) {
    var renderers = existingObjectRenderers.stream().filter(
        renderer -> renderer.supports(options)
    ).toList();

    if (renderers.isEmpty()) {
      throw new OptionsAreNotSupportedByAnyRendererException(options);
    }

    if (renderers.size() > 1) {
      throw new OptionsAreSupportedByMultipleRenderersException(options);
    }

    return renderers.stream()
        .findFirst()
        .orElseThrow();
  }


}
