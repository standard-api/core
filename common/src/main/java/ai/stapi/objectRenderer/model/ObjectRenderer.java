package ai.stapi.objectRenderer.model;

public interface ObjectRenderer {

  RenderOutput render(Object obj);

  RenderOutput render(Object obj, RendererOptions options);

  boolean supports(RendererOptions options);
}
