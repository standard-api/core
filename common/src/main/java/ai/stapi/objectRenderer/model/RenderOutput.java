package ai.stapi.objectRenderer.model;

public interface RenderOutput {

  String toPrintableString();
}
