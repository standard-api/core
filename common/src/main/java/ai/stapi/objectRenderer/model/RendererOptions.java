package ai.stapi.objectRenderer.model;

public interface RendererOptions {

}
