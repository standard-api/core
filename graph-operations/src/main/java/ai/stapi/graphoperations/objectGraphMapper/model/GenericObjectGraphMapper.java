package ai.stapi.graphoperations.objectGraphMapper.model;

import ai.stapi.graphoperations.graphLanguage.graphDescription.graphDescriptionBuilder.GraphDescriptionBuilder;
import ai.stapi.graphoperations.graphWriter.GenericGraphWriter;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectObjectGraphMapping;
import ai.stapi.graphoperations.objectGraphMapper.model.exceptions.ObjectGraphMapperException;
import ai.stapi.graphoperations.ogmProviders.GenericGraphMappingProvider;
import ai.stapi.serialization.SerializableObject;
import java.util.List;
import java.util.Map;

public class GenericObjectGraphMapper {

  private final List<SpecificObjectGraphMapper> specificObjectGraphMappers;
  private final GenericGraphWriter writer;
  private final GenericGraphMappingProvider mappingProvider;

  public GenericObjectGraphMapper(
      List<SpecificObjectGraphMapper> specificObjectGraphMappers,
      GenericGraphWriter writer,
      GenericGraphMappingProvider mappingProvider
  ) {
    this.specificObjectGraphMappers = specificObjectGraphMappers;
    this.writer = writer;
    this.mappingProvider = mappingProvider;
  }

  public GraphMappingResult mapToGraph(
      ObjectGraphMapping objectGraphMapping,
      Object object
  ) {
    return this.mapToGraph(objectGraphMapping, object, MissingFieldResolvingStrategy.STRICT);
  }

  public GraphMappingResult mapToGraph(
      ObjectGraphMapping objectGraphMapping,
      Object object,
      MissingFieldResolvingStrategy missingFieldResolvingStrategy
  ) {
    if (object == null) {
      throw ObjectGraphMapperException.becausegivenObjectIsNull(objectGraphMapping);
    }
    if (!(objectGraphMapping instanceof ObjectObjectGraphMapping)) {
      throw ObjectGraphMapperException.becauseFirstObjectGraphMappingIsNotObject(
          objectGraphMapping);
    }
    var builder = new GraphDescriptionBuilder();
    this.resolveInternally(
        objectGraphMapping,
        Map.entry("", object),
        builder,
        missingFieldResolvingStrategy
    );
    var graph = this.writer.createGraph(
        builder.getOnlyPositiveGraphDescriptions()
    );

    var elementsForRemoval = this.writer.createElementsForRemoval(
        builder.getOnlyRemovalGraphDescriptions()
    );
    return new GraphMappingResult(graph, elementsForRemoval);
  }

  public GraphMappingResult mapToGraph(SerializableObject object) {
    return this.mapToGraph(object, MissingFieldResolvingStrategy.STRICT);
  }

  public GraphMappingResult mapToGraph(
      SerializableObject object,
      MissingFieldResolvingStrategy missingFieldResolvingStrategy
  ) {
    var objectGraphMapping = this.mappingProvider.provideGraphMapping(object.getSerializationType());
    return this.mapToGraph(objectGraphMapping, object, missingFieldResolvingStrategy);
  }

  public GraphMappingResult mapToGraph(
      Object object,
      String serializationType,
      MissingFieldResolvingStrategy missingFieldResolvingStrategy
  ) {
    var objectGraphMapping = this.mappingProvider.provideGraphMapping(serializationType);
    return this.mapToGraph(objectGraphMapping, object, missingFieldResolvingStrategy);
  }

  public void resolveInternally(
      ObjectGraphMapping graphMapping,
      Map.Entry<String, Object> fieldEntry,
      GraphDescriptionBuilder builder,
      MissingFieldResolvingStrategy missingFieldResolvingStrategy
  ) {
    var supportingResolver = this.getSupportingSpecificMapper(graphMapping);
    supportingResolver.createGraphDescriptionWithValues(
        graphMapping,
        fieldEntry,
        builder,
        missingFieldResolvingStrategy
    );
  }

  private SpecificObjectGraphMapper getSupportingSpecificMapper(
      ObjectGraphMapping objectGraphMapping) {
    var supportingGraphWriters = this.specificObjectGraphMappers.stream()
        .filter(
            specificObjectGraphMapper -> specificObjectGraphMapper.supports(objectGraphMapping));

    var listOfSupportingResolvers = supportingGraphWriters.toList();
    if (listOfSupportingResolvers.size() == 0) {
      throw ObjectGraphMapperException.becauseNoSupportingSpecificResolverForGivenDeclaration(
          objectGraphMapping);
    }
    if (listOfSupportingResolvers.size() > 1) {
      throw ObjectGraphMapperException.becauseMoreThanOneSpecificResolverForGivenDeclaration(
          objectGraphMapping,
          listOfSupportingResolvers
      );
    }
    return listOfSupportingResolvers.get(0);
  }
}
