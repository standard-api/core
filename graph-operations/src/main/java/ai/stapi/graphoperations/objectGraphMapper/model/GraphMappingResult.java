package ai.stapi.graphoperations.objectGraphMapper.model;

import ai.stapi.graph.graphElementForRemoval.GraphElementForRemoval;
import ai.stapi.graph.Graph;
import java.util.ArrayList;
import java.util.List;

public class GraphMappingResult {

  private final List<GraphElementForRemoval> elementForRemoval;
  private final Graph graph;

  public GraphMappingResult(
      Graph graph,
      List<GraphElementForRemoval> elementForRemoval
  ) {
    this.elementForRemoval = elementForRemoval;
    this.graph = graph;
  }

  public List<GraphElementForRemoval> getElementForRemoval() {
    return elementForRemoval;
  }

  public Graph getGraph() {
    return graph;
  }

  public GraphMappingResult merge(GraphMappingResult anotherMappingResult) {
    var mergedList = new ArrayList<>(this.elementForRemoval);
    mergedList.addAll(anotherMappingResult.elementForRemoval);
    return new GraphMappingResult(
        this.getGraph().merge(anotherMappingResult.graph),
        mergedList
    );
  }
}
