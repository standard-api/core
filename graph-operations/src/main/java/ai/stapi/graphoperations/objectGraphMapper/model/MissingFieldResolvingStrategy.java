package ai.stapi.graphoperations.objectGraphMapper.model;

public enum MissingFieldResolvingStrategy {

  STRICT,
  LENIENT
}
