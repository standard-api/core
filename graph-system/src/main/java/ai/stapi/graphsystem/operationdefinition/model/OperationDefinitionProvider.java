package ai.stapi.graphsystem.operationdefinition.model;

import ai.stapi.graphsystem.operationdefinition.exceptions.CannotProvideOperationDefinition;
import java.util.List;

public interface OperationDefinitionProvider {

  List<OperationDefinitionDTO> provideAll();

  OperationDefinitionDTO provide(String operationId) throws CannotProvideOperationDefinition;

  boolean contains(String operationId);
}
