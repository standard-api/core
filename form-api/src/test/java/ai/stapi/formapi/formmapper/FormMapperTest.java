package ai.stapi.formapi.formmapper;

import ai.stapi.formapi.formmapper.fixtures.FormApiTestDefinitionsLoader;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionProvider;
import ai.stapi.test.schemaintegration.SchemaIntegrationTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope(FormApiTestDefinitionsLoader.SCOPE)
class FormMapperTest extends SchemaIntegrationTestCase {
  
  @Autowired
  private OperationDefinitionProvider operationDefinitionProvider;
  
  @Autowired
  private FormMapper formMapper;

  @Test
  void itShouldReturnEmptySchemaForOperationDefinitionWithoutParameters() {
    var operationDefinition = this.operationDefinitionProvider.provide("TestEmptyCommand");
    var formSchema = this.formMapper.map(operationDefinition);
    this.thenObjectApproved(formSchema);
  }

  @Test
  void itShouldReturnSchemaForOperationDefinitionWithStringParameter() {
    var operationDefinition = this.operationDefinitionProvider.provide("TestSimpleCommand");
    var formSchema = this.formMapper.map(operationDefinition);
    this.thenObjectApproved(formSchema);
  }

}