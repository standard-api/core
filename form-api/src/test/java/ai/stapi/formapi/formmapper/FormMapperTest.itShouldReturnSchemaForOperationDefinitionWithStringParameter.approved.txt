{
  "type" : "object",
  "properties" : {
    "exampleStringField" : {
      "type" : "string"
    }
  },
  "required" : [ "exampleStringField" ]
}